### PR TITLE
Partial structure defined

### PR DESCRIPTION
**Pull request by sjmog**
OK gang, _what do we think_? Please **read this before checking code**

Recap of the main reasoning:

1. Non-developers can use an easy-to-use DSL to build pages using set molecules
2. Developers _rarely_ have to CSS
3. It's suitable for moving to a form-based, then later drag-and-drop-based, interface builder.

### Top Level
Non-developers can build interfaces using a very easy syntax. Here is an example (the 'meet our grads' block) from `homepage.erb`:
```erb
<!-- BEGIN grads -->

<%= partial "components/molecules/media-component--with-slider", locals: {
    leader: "Meet",
    title: "Our Grads",
    body_copy: "We&rsquo;ve graduated over 300 beginners into their dream jobs as junior developers. Some were looking for a career change, others to define themselves.",
    cta: {
      text: "Read Grad Stories",
      link: "/graduates",
      type: "horizontal"
    },
    slider_content: [
      {
        content_type: "image",
        src: "images/test-assets/test-grad.png",
        orientation: "portrait",
        alt: "A test graduate, looking at the camera.",
        caption: "Firstname Lastname\nCohort of 2015\nNow a Junior Developer at CompanyName"
      },
      {
        content_type: "image",
        src: "http://placehold.it/400x600&text=400x600",
        orientation: "portrait",
        alt: "A gray box with the text '400x600'",
        caption: "Othername Lastname\nCohort of 2015\nNow a Junior Developer at OtherCompanyName"
      }
    ]
  } %>

<!-- END grads -->
```
_Molecule partials are very specific_. When a user renders a partial called `media-component--with-slider`, they know exactly what they're getting, and there is a standard set of attributes for them to fill in.

### Bottom level
Developers can build new molecules using atoms. These atoms are very generic. Here is the `cta-block` atom, which renders any number of Calls to Action (buttons) in a supplied configuration:
```erb
<% # Includes any number of _cta quarks %>
<% component ||= "cta-block" %>

<div class="<%= component_classes locals.fetch(:parent, nil), component, locals.fetch(:modifiers, []) %>">

  <% ctas.each do | cta | %>
    <%= partial "components/quarks/cta", locals: {
      parent: component,
      text: cta[:text],
      link: cta[:link],
      modifiers: cta[:modifiers]
    } %>
  <% end %>

</div>
```

Any component (except molecule components) is rendered by including **two main attributes**:

1. The component's `parent` (a string), and
2. Any `modifiers` (an array of strings).

So, to render the `cta-block` above:

```erb
<!-- this will render the 'cta-block' atom, with a single 'cta' quark inside -->
<%= partial "components/atoms/cta-block", locals: {
  parent: "media-component__section", # If the parent is a media-component__section
  modifiers: ["left"], # Or any other modifiers
  # the following is an array of cta quarks (only one, here)
  ctas: [
    { 
      text: cta.fetch(:text, "Untitled"),
      link: cta.fetch(:link, "#"),
      modifiers: [cta.fetch(:type, "horizontal")]
    }
  ]
} %>
```
If you provide a `parent` and a `modifiers` array, the builder will include the correct BEM classes for you (see the `ComponentClassesHelper` in `lib`). If you don't include these, the BEM helper will still work and you will render the most generic form of a component.

_Atom and Quark partials are very generic_. They accept a wide range of (sensibly-named) attributes, and my next task will be to document them.

### Between Molecules and other Components
Molecule partials (go check out `components/molecules/_media-component--with-slider.erb`) should:

1. Translate between the top-level, 'non-dev' attributes (which are concrete and easy to understand), and the more abstract component attributes detailed above, and 
2. Provide defaults should a 'non-dev' forget to include an attribute when building a page.

OK, I think that's enough. If there's anything here that doesn't make sense, **please grab me**. As Roi will testify, this is easier to get your head round than it might seem, and it hits each point of the main reasoning well.

### CODE REVIEW

**Comment by pitchinvasion** [comment]
We've discussed and are all happy (as is Arfa) , although we're going to change the names to atoms -> groups -> templates. Me, Roi and Steve are going to work renaming and adding documentation and generally start turning this into something we can prototype for the beta site.

Awesome work Sam thanks for getting this done

File: spec/makers_styles/helpers/caption_newlines_to_caption_lines_helper_spec.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
+require 'makers_styles/helpers/caption_newlines_to_caption_bodies_helper'
+
+class CaptionNewlinesToCaptionBodiesHelperWrapper
+  include CaptionNewlinesToCaptionBodiesHelper
+end
+
+describe CaptionNewlinesToCaptionBodiesHelperWrapper do
+  it "translates newline entities to an array of hashes" do
```
Comment (pitchinvasion) [inline spec/makers_styles/helpers/caption_newlines_to_caption_lines_helper_spec.rb]: I tend to focus on behaviour in unit tests, whereas this is looking at something more code-related - i.e. that this returns a specific type. In Ruby we tend to follow duck typing conventions so we aren't bothered what the type of something is normally. You can probably delete this test as anyway the next test covers this test as well

File: source/components/atoms/_image-block.erb [hunk 2]
```diff
@@ -0,0 +1,22 @@
+<% component ||= "image-block" %>
```
Comment (pitchinvasion) [inline source/components/atoms/_image-block.erb]: How come this bit doesn't use fetch like the other locals? Also is there an occasion when the component wouldn't be an image block?

File: lib/makers_styles/helpers/caption_newlines_to_caption_bodies_helper.rb [hunk 3]
```diff
@@ -0,0 +1,5 @@
+module CaptionNewlinesToCaptionBodiesHelper
+  def newlines_to_caption_bodies caption
```
Comment (pitchinvasion) [inline lib/makers_styles/helpers/caption_newlines_to_caption_bodies_helper.rb]: I could see this being quite a common theme where we want to pass multiple paragraphs - I would suggest it might be cleaner to do it the Rails way, which is to use a generic helper method called `auto_p` that translates new lines into <p> tags. This would take out some of the complexity in the caption atom for example as we wouldn't need the `body` atom which is just really a <p> tag

Comment (pitchinvasion) [inline lib/makers_styles/helpers/caption_newlines_to_caption_bodies_helper.rb]: Oh wait every paragraph has BEM classes at the moment - is that neccesary in BEM? I guess if it is we'll have to use a helper, but is this not overkill just to add a paragraph of text?

File: lib/makers_styles/helpers/component_classes_helper.rb [hunk 4]
```diff
@@ -0,0 +1,12 @@
+module ComponentClassesHelper
+  def component_modifiers component, modifiers
+    modifiers.inject("#{component}") { | str, mod | str << " #{component}--#{mod}" }
+  end
+
+  def component_classes parent=nil, component="", modifiers=[]
+    modifiers ||= []
```
Comment (spike01) [inline lib/makers_styles/helpers/component_classes_helper.rb]: This is covered by the default arg in the line above, no?

File: lib/makers_styles/helpers/component_classes_helper.rb [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+    modifier_classes = modifiers.inject("#{component}") { | str, mod | str << " #{component}--#{mod}" }
```
Comment (spike01) [inline lib/makers_styles/helpers/component_classes_helper.rb]: Is this not `#component_modifiers` above?

File: spec/makers_styles/helpers/component_modifiers_helper_spec.rb [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+  end
+
+  context "in the absence of both parent components and modifiers" do
+    it "constructs component classes without modifier or block-level classes" do
+      parent = nil
+      component = "cta"
+      modifiers = []
+      expect(subject.component_classes(parent, component, modifiers)).to eq "cta"
+    end
+  end
```
Comment (spike01) [inline spec/makers_styles/helpers/component_modifiers_helper_spec.rb]: The two tests above don't really make sense unless lines 28 and 38 are deleted?

**Comment by spike01** [comment]
I'm just being nitpicky with my comments! Otherwise I actually quite like where this is going.

File: lib/makers_styles/helpers/component_classes_helper.rb [hunk 7]
```diff
@@ -0,0 +1,12 @@
+module ComponentClassesHelper
+  def component_modifiers component, modifiers
+    modifiers.inject("#{component}") { | str, mod | str << " #{component}--#{mod}" }
+  end
```
Comment (spike01) [inline lib/makers_styles/helpers/component_classes_helper.rb]: This method isn't called anywhere, nor tested

File: spec/makers_styles/helpers/caption_newlines_to_caption_lines_helper_spec.rb (hunk 1 above)
Comment (sjmog) [inline spec/makers_styles/helpers/caption_newlines_to_caption_lines_helper_spec.rb]: Good call

File: lib/makers_styles/helpers/component_classes_helper.rb (hunk 7 above)
Comment (sjmog) [inline lib/makers_styles/helpers/component_classes_helper.rb]: It's called in pretty much every partial

It's tested in `spec/makers_styles/helpers/component_classes_helper.rb`

File: source/components/atoms/_image-block.erb (hunk 2 above)
Comment (sjmog) [inline source/components/atoms/_image-block.erb]: Literally no reason I think

File: lib/makers_styles/helpers/caption_newlines_to_caption_bodies_helper.rb (hunk 3 above)
Comment (sjmog) [inline lib/makers_styles/helpers/caption_newlines_to_caption_bodies_helper.rb]: This is actually to format the supplied `caption` into something the `_caption.erb` atom will take.

See the test for more -> but it doesn't just turn it into `<p>`s

`_caption.erb` expects an array of `bodies`, whose elements are Hashes containing a `text` key (so it can render multi-line bodies: this is a rare example where one atom contains another).

`newlines_to_caption_bodies` takes a sentence like `Lorem\nIpsum` and turns it into

```ruby
[{ text: "Lorem" }, { text: "Ipsum" }]
```

which can then be fed into `bodies` for the `_caption.erb` to render on new lines.

File: lib/makers_styles/helpers/component_classes_helper.rb (hunk 4 above)
Comment (sjmog) [inline lib/makers_styles/helpers/component_classes_helper.rb]: Yes - this is left over from an earlier modification, can kill

File: lib/makers_styles/helpers/component_classes_helper.rb (hunk 5 above)
Comment (sjmog) [inline lib/makers_styles/helpers/component_classes_helper.rb]: Yes -> I meant to put component modifiers into a private method here

File: spec/makers_styles/helpers/component_modifiers_helper_spec.rb (hunk 6 above)
Comment (sjmog) [inline spec/makers_styles/helpers/component_modifiers_helper_spec.rb]: How so? When modifiers is undefined, we get horrible things happening, so modifiers defaults to n empty array (in the views, using `fetch` on the provided `locals`)

File: lib/makers_styles/helpers/component_classes_helper.rb (hunk 4 above)
Comment (sjmog) [inline lib/makers_styles/helpers/component_classes_helper.rb]: Build a bunch o private methods to make it prettier

File: source/components/molecules/_media-component--with-slider.erb [hunk 8]
```diff
@@ -0,0 +1,68 @@
+<% component = "media-component" %>
+<% modifiers = locals.fetch(:modifiers, []) %>
+<% alignment_modifiers = modifiers.include?("reverse") ? ["right"] : ["left"] %>
```
Comment (sjmog) [inline source/components/molecules/_media-component--with-slider.erb]: This is where I feel uneasy, crew: there's some logic (albeit view/presentational logic) in the view. There'll be more too, the more variants are built. What do we think about this?

File: source/components/molecules/_media-component--with-slider.erb [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+                    src: slider_item.fetch(:src, "http://placehold.it/400x600"),
+                    alt: slider_item.fetch(:alt, "Untitled Image")
+                  }
+                ],
+                caption: {
+                  modifiers: ["right"],
+                  bodies: newlines_to_caption_bodies(slider_item.fetch(:caption, ""))
+                }
+              } %>
+          <% elsif slider_item[:content_type] === "title-block" %>
```
Comment (sjmog) [inline source/components/molecules/_media-component--with-slider.erb]: This must be constructed next --> that way, any sort of content can be provided to a `media-block--with-slider` and chucked straight into the slider

Other than images and title blocks, I can't think of what else might be passed this way right now

File: source/components/quarks/_body.erb [hunk 10]
```diff
@@ -0,0 +1,5 @@
+<% component = "body" %>
+
+<p class="<%= component_classes locals.fetch(:parent, nil), component, locals.fetch(:modifiers, []) %>">
```
Comment (sjmog) [inline source/components/quarks/_body.erb]: Can we think of a neater way to do the `component_classes` helper? I'm repeating `locals.fetch` a **lot** right now

File: source/homepage.erb [hunk 11]
```diff
@@ -56,52 +56,32 @@
...
+            src: "images/test-assets/test-grad.png",
+            orientation: "portrait",
+            alt: "A test graduate, looking at the camera.",
+            caption: "Firstname Lastname\nCohort of 2015\nNow a Junior Developer at CompanyName"
+          },
+          {
+            content_type: "image",
+            src: "http://placehold.it/400x600&text=400x600",
+            orientation: "portrait",
+            alt: "A gray box with the text '400x600'",
```
Comment (sjmog) [inline source/homepage.erb]: Roi and I made the decision to require an `alt` tag on images --> good accessibility. What does everyone think of this?

File: source/homepage.erb [hunk 12]
```diff
@@ -152,6 +132,32 @@
...
+          {
+            content_type: "image",
+            src: "images/test-assets/test-curric.jpg",
+            orientation: "portrait",
+            alt: "Rails is a web application framework for large systems."
+          }
+        ]
+      } %>
+
     <div class="media-component media-component--reverse">
```
Comment (sjmog) [inline source/homepage.erb]: The above partial is repeated here, because the partial is not entirely complete (I was constructing it to demo to Steve)

File: source/index.erb [hunk 13]
```diff
@@ -156,6 +156,7 @@ title: MA Style Guide
     <li><%= link_to "Inputs", "visual_guide/quarks/input.html" %></li>
     <li><%= link_to "Images", "visual_guide/quarks/image.html" %></li>
     <li><%= link_to "Captions", "visual_guide/quarks/caption.html" %></li>
+    <li><%= link_to "Body", "visual_guide/quarks/body.html" %></li>
```
Comment (sjmog) [inline source/index.erb]: This should form the basis of the visual guide

File: source/components/molecules/_media-component--with-slider.erb (hunk 8 above)
Comment (pitchinvasion) [inline source/components/molecules/_media-component--with-slider.erb]: why does reverse need to be included? can't you just pass right or left as a modifier?

File: source/components/quarks/_body.erb (hunk 10 above)
Comment (pitchinvasion) [inline source/components/quarks/_body.erb]: yes just pass it locals, and let `component_classes` handle doing the fetching

File: source/homepage.erb (hunk 11 above)
Comment (pitchinvasion) [inline source/homepage.erb]: Yep fine, good for SEO too

File: source/components/molecules/_media-component--with-slider.erb (hunk 8 above)
Comment (sjmog) [inline source/components/molecules/_media-component--with-slider.erb]: Steve asked this yesterday: I was trying to avoid explicitly linking Template names to presentation like that (not so fussed at the Group level)

File: spec/makers_styles/helpers/component_modifiers_helper_spec.rb (hunk 6 above)
Comment (spike01) [inline spec/makers_styles/helpers/component_modifiers_helper_spec.rb]: I thought that `component_classes` catches modifiers being `nil` through its default arguments... i.e. when you don't provide a modifier in your view, you're not supplying empty array, you're actually providing `nil`? That's what I got from my cursory reading...

File: lib/makers_styles/helpers/component_classes_helper.rb (hunk 7 above)
Comment (spike01) [inline lib/makers_styles/helpers/component_classes_helper.rb]: Indeed, or inlined with the method underneath? Since it kinda is already

File: source/components/molecules/_media-component--with-slider.erb (hunk 8 above)
Comment (pitchinvasion) [inline source/components/molecules/_media-component--with-slider.erb]: not entirely clear to me why that's a huge issue, but if it is we could use a helper

File: source/homepage.erb (hunk 11 above)
Comment (roidriscoll) [inline source/homepage.erb]: I'm fine with it :wink:

File: source/components/molecules/_media-component--with-slider.erb (hunk 8 above)
Comment (spike01) [inline source/components/molecules/_media-component--with-slider.erb]: Would make more sense as a helper, like that if there are any changes it happens in only one place, not in every single view

File: source/components/quarks/_body.erb (hunk 10 above)
Comment (spike01) [inline source/components/quarks/_body.erb]: Can you pass all of `locals` into `component_classes` and do the destructuring in the helper?

File: spec/makers_styles/helpers/component_modifiers_helper_spec.rb (hunk 6 above)
Comment (sjmog) [inline spec/makers_styles/helpers/component_modifiers_helper_spec.rb]: Nope - when you don't provide a `modifiers` array _to_ your view, it defaults to providing `[]` to this method

File: source/components/molecules/_media-component--with-slider.erb (hunk 8 above)
Comment (sjmog) [inline source/components/molecules/_media-component--with-slider.erb]: This is an excellent plan: for this PR I only schedaddled up one Template, then Steve asked to take a peek so I shovelled in another. How we handle Template layout modifiers is up for debate

e.g. there is a component, `media-component--with-slider` with the following modifiers:

- 'regular' (with the slider on the right)
- 'reverse' (with the slider on the left)
- 'gs' (with one section at 61.8% width and the other at 38.2%)

These modifiers can be mixed (so a user could supply a `["reverse", "gs"]` modifiers local to the Template).

I believe we should handle all three of the above in a single Template, and roll up a completely different template for, e.g., a `media-component--with-form` (which may also have different modifiers). 

**TL;DR We need to decide on a formal limit for what requires a modifier, and what requires a completely new Template.**

File: source/components/quarks/_body.erb (hunk 10 above)
Comment (sjmog) [inline source/components/quarks/_body.erb]: **Obviously** that makes an enormous amount more sense, no idea why I failed to spot that duhhh
